File: samples/cpp/backup.cpp
/*
 *  backup.cpp — companion sample for ../../backup-and-recovery.md
 *
 *  A gbak backup + restore round trip driven entirely through the Services
 *  API — the same path `gbak -se` and fbsvcmgr use, so no gbak binary is
 *  needed on the client:
 *
 *    1. create a scratch database with a table and a few rows;
 *    2. attach to service_mgr and start isc_action_svc_backup (verbose),
 *       streaming gbak's log line by line with isc_info_svc_line;
 *    3. start isc_action_svc_restore (replace) the same way;
 *    4. attach to the restored database and prove the rows survived.
 *
 *  The backup runs while the source attachment is still open: gbak reads
 *  through a snapshot transaction, the "Online" property of the document's
 *  gbak section.
 */

#include "fb_sample.h"

using namespace Firebird;
using fbsample::master;

static const char* DB_SRC  = "inet://localhost//tmp/fbhandson/backup.fdb";
static const char* SRV     = "localhost:service_mgr";
static const char* FBK     = "/tmp/fbhandson/backup.fbk";           // server-side path
static const char* DB_REST = "inet://localhost//tmp/fbhandson/backup_restored.fdb";

// Drain one service's output: query isc_info_svc_line until the line is empty.
static void streamServiceOutput(ThrowStatusWrapper& st, IService* svc, const char* pfx)
{
	const unsigned char items[] = { isc_info_svc_line };
	unsigned char buf[1024];
	for (;;)
	{
		svc->query(&st, 0, nullptr, sizeof(items), items, sizeof(buf), buf);
		if (buf[0] != isc_info_svc_line)
			break;
		const unsigned len = buf[1] | (buf[2] << 8);        // little-endian VAX word
		if (!len)
			break;                                          // empty line = service done
		printf("%s%.*s\n", pfx, len, buf + 3);
	}
}

int main(int argc, char** argv)
{
	try
	{
		// -- 1. scratch source database ---------------------------------
		fbsample::Db db;
		db.attachOrCreate(fbsample::argOrDefault(argc, argv, 1, DB_SRC));
		ITransaction* tra = db.start();
		try { db.exec(tra, "DROP TABLE BR_ITEMS"); } catch (const FbException&) {}
		db.exec(tra, "CREATE TABLE BR_ITEMS (ID INT NOT NULL PRIMARY KEY, NAME VARCHAR(30))");
		tra->commitRetaining(&db.status);
		db.exec(tra, "INSERT INTO BR_ITEMS VALUES (1, 'alpha')");
		db.exec(tra, "INSERT INTO BR_ITEMS VALUES (2, 'beta')");
		db.exec(tra, "INSERT INTO BR_ITEMS VALUES (3, 'gamma')");
		tra->commit(&db.status);
		printf("source ready: BR_ITEMS with 3 rows\n");

		// -- 2. attach to the service manager ---------------------------
		ThrowStatusWrapper st(master->getStatus());
		IProvider* prov = master->getDispatcher();
		IUtil* utl = master->getUtilInterface();

		IXpbBuilder* spb = utl->getXpbBuilder(&st, IXpbBuilder::SPB_ATTACH, nullptr, 0);
		spb->insertString(&st, isc_spb_user_name, fbsample::Db::defaultUser());
		spb->insertString(&st, isc_spb_password, fbsample::Db::defaultPassword());
		IService* svc = prov->attachServiceManager(&st, SRV,
			spb->getBufferLength(&st), spb->getBuffer(&st));
		spb->dispose();

		// -- 3. gbak backup through the service, verbose ----------------
		printf("\n== backup: %s -> %s ==\n", "/tmp/fbhandson/backup.fdb", FBK);
		IXpbBuilder* start = utl->getXpbBuilder(&st, IXpbBuilder::SPB_START, nullptr, 0);
		start->insertTag(&st, isc_action_svc_backup);
		start->insertString(&st, isc_spb_dbname, "/tmp/fbhandson/backup.fdb");
		start->insertString(&st, isc_spb_bkp_file, FBK);
		start->insertTag(&st, isc_spb_verbose);
		svc->start(&st, start->getBufferLength(&st), start->getBuffer(&st));
		start->dispose();
		streamServiceOutput(st, svc, "  gbak> ");

		// -- 4. gbak restore (replace) through the same service ---------
		printf("\n== restore: %s -> %s ==\n", FBK, "/tmp/fbhandson/backup_restored.fdb");
		start = utl->getXpbBuilder(&st, IXpbBuilder::SPB_START, nullptr, 0);
		start->insertTag(&st, isc_action_svc_restore);
		start->insertString(&st, isc_spb_bkp_file, FBK);
		start->insertString(&st, isc_spb_dbname, "/tmp/fbhandson/backup_restored.fdb");
		start->insertInt(&st, isc_spb_options, isc_spb_res_replace);
		start->insertTag(&st, isc_spb_verbose);
		svc->start(&st, start->getBufferLength(&st), start->getBuffer(&st));
		start->dispose();
		streamServiceOutput(st, svc, "  gbak> ");

		svc->detach(&st);
		prov->release();
		st.dispose();

		// -- 5. prove the restored copy has the data --------------------
		fbsample::Db rdb;
		rdb.attach(DB_REST);
		ITransaction* rtra = rdb.start();
		printf("\nrestored database says: %s rows, max name = %s\n",
			rdb.queryValue(rtra, "SELECT COUNT(*) FROM BR_ITEMS").c_str(),
			rdb.queryValue(rtra, "SELECT MAX(NAME) FROM BR_ITEMS").c_str());
		rtra->commit(&rdb.status);
		printf("done.\n");
		return 0;
	}
	catch (const FbException& e)
	{
		return fbsample::report(e);
	}
}

File: samples/fpc/backup.pas
{
  backup.pas — a gbak backup + restore round trip through the Services API.

  The fbintf twin of ../cpp/backup.cpp, companion to
  ../../backup-and-recovery.md.  The same path `gbak -se` and fbsvcmgr use,
  no gbak binary needed on the client:

    1. create a scratch database with a table and a few rows;
    2. IServiceManager + isc_action_svc_backup (verbose), streaming gbak's
       log line by line with isc_info_svc_line;
    3. isc_action_svc_restore (replace) the same way;
    4. attach to the restored database and prove the rows survived.

  The backup runs while the source attachment is still open: gbak reads
  through a snapshot transaction — the "online" property of the document's
  gbak section.  Driver contrast: fbintf reaches the Services API natively
  (IServiceManager, IB.pas) just like the C++ OO API and node-firebird's
  pure-JS client; rsfbclient cannot, and its twin shells out to gbak.

  Build & run (see ../README.md):
      make -C samples/fpc bin/backup && samples/fpc/bin/backup
}
program backup;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses SysUtils, IB, FBHandsOn;

const FBK = '/tmp/fbhandson/backup_fpc.fbk';               {server-side path}
      DB_REST = '/tmp/fbhandson/backup_fpc_restored.fdb';

{Drain one service's verbose output: query isc_info_svc_line until empty.}
procedure StreamServiceOutput(Svc: IServiceManager; const pfx: AnsiString);
var Req: ISRB;
    Res: IServiceQueryResults;
    Item: IServiceQueryResultItem;
    line: AnsiString;
begin
  repeat
    Req := Svc.AllocateSRB;
    Req.Add(isc_info_svc_line);
    Res := Svc.Query(Req);
    Item := Res.Find(isc_info_svc_line);
    if Item = nil then break;
    line := Item.AsString;                 {empty line = service done}
    if line <> '' then
      writeln(pfx, line);
  until line = '';
end;

var Att, RAtt: IAttachment;
    Tr: ITransaction;
    SPB: ISPB;
    Svc: IServiceManager;
    Req: ISRB;

begin
  { -- 1. scratch source database --------------------------------- }
  Att := AttachOrCreate(DbConn('backup'));
  Tr := Att.StartTransaction([isc_tpb_write, isc_tpb_nowait,
    isc_tpb_concurrency], taCommit);
  try
    Att.ExecuteSQL(Tr, 'DROP TABLE BR_ITEMS', []);
  except on EIBInterBaseError do ;
  end;
  Att.ExecuteSQL(Tr,
    'CREATE TABLE BR_ITEMS (ID INT NOT NULL PRIMARY KEY, NAME VARCHAR(30))', []);
  Tr.CommitRetaining;
  Att.ExecuteSQL(Tr, 'INSERT INTO BR_ITEMS VALUES (1, ''alpha'')', []);
  Att.ExecuteSQL(Tr, 'INSERT INTO BR_ITEMS VALUES (2, ''beta'')', []);
  Att.ExecuteSQL(Tr, 'INSERT INTO BR_ITEMS VALUES (3, ''gamma'')', []);
  Tr.Commit;
  writeln('source ready: BR_ITEMS with 3 rows');

  { -- 2. attach to the service manager --------------------------- }
  SPB := FirebirdAPI.AllocateSPB;
  SPB.Add(isc_spb_user_name).AsString := HandsOnUser;
  SPB.Add(isc_spb_password).AsString := HandsOnPassword;
  Svc := FirebirdAPI.GetServiceManager('localhost', TCP, SPB);

  { -- 3. gbak backup through the service, verbose ---------------- }
  writeln;
  writeln('== backup: ', DbPath('backup'), ' -> ', FBK, ' ==');
  Req := Svc.AllocateSRB;
  Req.Add(isc_action_svc_backup);
  Req.Add(isc_spb_dbname).AsString := DbPath('backup');
  Req.Add(isc_spb_bkp_file).AsString := FBK;
  Req.Add(isc_spb_verbose);
  Svc.Start(Req);
  StreamServiceOutput(Svc, '  gbak> ');

  { -- 4. gbak restore (replace) through the same service --------- }
  writeln;
  writeln('== restore: ', FBK, ' -> ', DB_REST, ' ==');
  Req := Svc.AllocateSRB;
  Req.Add(isc_action_svc_restore);
  Req.Add(isc_spb_bkp_file).AsString := FBK;
  Req.Add(isc_spb_dbname).AsString := DB_REST;
  Req.Add(isc_spb_options).SetAsInteger(isc_spb_res_replace);
  Req.Add(isc_spb_verbose);
  Svc.Start(Req);
  StreamServiceOutput(Svc, '  gbak> ');
  Svc.Detach;

  { -- 5. prove the restored copy has the data -------------------- }
  RAtt := FirebirdAPI.OpenDatabase('localhost:' + DB_REST, DefaultDPB);
  Tr := RAtt.StartTransaction([isc_tpb_read, isc_tpb_nowait,
    isc_tpb_concurrency], taCommit);
  writeln;
  writeln('restored database says: ',
    RAtt.OpenCursorAtStart(Tr, 'SELECT COUNT(*) FROM BR_ITEMS')[0].AsString,
    ' rows, max name = ',
    RAtt.OpenCursorAtStart(Tr, 'SELECT MAX(NAME) FROM BR_ITEMS')[0].AsString);
  Tr.Commit;
  writeln('(Services API native in fbintf - node-firebird reimplements it in JS;');
  writeln(' rsfbclient has no services attach and must shell out to gbak.)');
  writeln('done.');
end.
